src/QMCWaveFunctions/Jastrow/JeeIOrbitalSoA.h

```diff
@@ -969,6 +986,106 @@ class JeeIOrbitalSoA : public WaveFunctionComponent
+  {
+    ParticleSet::ParticleGradient_t tempG;
+    ParticleSet::ParticleLaplacian_t tempL;
+    tempG.resize(P.getTotalNum());
+    tempL.resize(P.getTotalNum());
+    QTFull::RealType delta = 0.00001;
+    QTFull::RealType c1   = 1.0 / delta / 2.0;
+    QTFull::RealType c2   = 1.0 / delta / delta;
+
+    GradType g_return(0.0);
+    // GRAD TEST COMPUTATION
+    PosType rI = source.R[isrc];
+    for (int iondim = 0; iondim < 3; iondim++)
+    {
+      source.R[isrc][iondim] = rI[iondim] + delta;
+      source.update();
+      P.update();
+
+      LogValueType log_p = evaluateLog(P, tempG, tempL);
+
+      source.R[isrc][iondim] = rI[iondim] - delta;
+      source.update();
+      P.update();
+      LogValueType log_m = evaluateLog(P, tempG, tempL);
+
+      QTFull::RealType log_p_r(0.0), log_m_r(0.0);
+      
+      log_p_r=log_p.real();
+      log_m_r=log_m.real();
+      //symmetric finite difference formula for gradient.
+      g_return[iondim] = c1 * (log_p_r - log_m_r);
+
+      //reset everything to how it was.
+      source.R[isrc][iondim] = rI[iondim];
+    }
+    // this last one makes sure the distance tables and internal neighbourlist correspond to unperturbed source.
+    source.update();
+    P.update();
+    build_compact_list(P);
+    return g_return;
+  }
+
+  inline GradType evalGradSource(ParticleSet& P,
+                                 ParticleSet& source,
+                                 int isrc,
+                                 TinyVector<ParticleSet::ParticleGradient_t, OHMMS_DIM>& grad_grad,
+                                 TinyVector<ParticleSet::ParticleLaplacian_t, OHMMS_DIM>& lapl_grad)
+  {
+    ParticleSet::ParticleGradient_t Gp, Gm, dG;
+    ParticleSet::ParticleLaplacian_t Lp, Lm, dL;
+    Gp.resize(P.getTotalNum());
+    Gm.resize(P.getTotalNum());
+    dG.resize(P.getTotalNum());
+    Lp.resize(P.getTotalNum());
+    Lm.resize(P.getTotalNum());
+    dL.resize(P.getTotalNum());
+
+    QTFull::RealType delta = 0.00001;
+    QTFull::RealType c1   = 1.0 / delta / 2.0;
+    QTFull::RealType c2   = 1.0 / delta / delta; 
+    GradType g_return(0.0);
+    // GRAD TEST COMPUTATION
+    PosType rI = source.R[isrc];
+    for (int iondim = 0; iondim < 3; iondim++)
+    {
+      Lp                     = 0;
+      Gp                     = 0;
+      Lm                     = 0;
+      Gm                     = 0;
+      source.R[isrc][iondim] = rI[iondim] + delta;
+      source.update();
+      P.update();
+
+      LogValueType log_p = evaluateLog(P, Gp, Lp);
+
+      source.R[isrc][iondim] = rI[iondim] - delta;
+      source.update();
+      P.update();
+      LogValueType log_m = evaluateLog(P, Gm, Lm);
+      QTFull::RealType log_p_r(0.0), log_m_r(0.0);
+      
+      log_p_r=log_p.real();
+      log_m_r=log_m.real();
+      dG=Gp-Gm;
+      dL=Lp-Lm;
+      //symmetric finite difference formula for gradient.
+      g_return[iondim] = c1 * (log_p_r - log_m_r);
```

Comment:
This formulation is only correct without node-crossing. I noticed this pattern in wftest and we need certain protection there as well.

Reply:
OK. It is fine here since Jastrow is bosonic. But not a general case in wftest.

Reply:
What are you thinking?  Check for a sign difference and abort if that happens?  Abort if log_p or log_m gets really really negative?

Reply:
It's moot for the Jastrow, but a valid concern for wftester that I hadn't considered.

Reply:
No problem here in the Jastrow.
In the generic case, the sign of psi matters and so as the phase of log(psi) not the sign of log(psi).

Reply:
Gotcha.